### DISCUSSION TITLE
Allow framework components to be pages

### DISCUSSION THREAD

**Discussion by matthewp** [withastro/roadmap#750] [Proposal]
### Body

# Summary

Allow any component backed by a renderer (what we usually call "framework components") to be used as a page and allowed in `src/pages/`.

# Background & Motivation

Astro components were created in order to allow the island pattern. Over time they've become more featureful and liked by the community, but there is always a context switch required when you go from working on an Astro component to work in a framework component for an island.

Additionally some people do not like the Astro syntax for subjective reasons. Since Astro already is agnostic about what framework you use, it seems reasonable to extend this to allow the same framework components to become pages.

# Goals

- Make it possible for any registered framework component to be used inside of `src/pages/`.
- Provide an API so that islands can be defined inside of the framework component.
- Provide an API for Astro's "global" object that can be provided to framework components.

# Example

Unknown API at this time, plan on updating this with discussion.

**Comment by lilnasy** [comment]
I am concerned it strays Astro's image and maintenance efforts sideways much more than it progresses them forward. My initial reaction is highly skeptical.

I can see a couple of APIs:
- similar to Remix - `export function frontmatter(Astro: AstroGlobal): Promise<{ props: any, client: "load" | "idle" | ..., prerender?: boolean  }>`
- similar to astro's compiler output - `export default function (Astro: AstroGlobal): Promise<FrameworkComponentInstance>`

...but whatever the API, there are going to be multiple ways to do the same thing. This is going to cause confusion and a harmful perception of complexity and bloat. "Which way is the recommended one?" "Is the old way deprecated?" "Ugh I just migrated to 3.x". Ultimately, it's a DX loss.

I understand that the proposal may be barebones just to get the conversation started, and my initial reaction may not be valid, but this is a drastic change. I would expect the problem it solves to be drastic as well. What are the wins here?

**Reply by a-y-u-s-h** [reply]
As a new Astro user, I was confused when the promise was "Bring your own framework", which insinuates that I can bring my framework anywhere, including the `pages` directory. Then I found out that I need to have certain `.astro` files in the project + I can have my components elsewhere with `client:load` directive, except the state management doesn't quite work the same way (like I can't use Contexts in React). I think the promise of "Bring your own framework" is best kept when you actually allow it everywhere, including allowing methods of state management in those frameworks (enable/disable islands on certain pages so that migration can happen easily?). I'm not sure what I actually want to be honest, but just wanted to put these points across from a beginner's perspective.

**Comment by louiss0** [comment]
I never thought this would be a thing but I hope it does. It would make the framework more competitive. 

There are challenges to this like are pages server components or can the user choose.
